tephra-core/src/main/java/co/cask/tephra/TransactionManager.java

```diff
@@ -470,12 +479,34 @@ private void restoreSnapshot(TransactionSnapshot snapshot) {
     readPointer = snapshot.getReadPointer();
     lastWritePointer = snapshot.getWritePointer();
     invalid.addAll(snapshot.getInvalid());
-    inProgress.putAll(snapshot.getInProgress());
+    inProgress.putAll(txnBackwardsCompatCheck(snapshot.getInProgress()));
     committingChangeSets.putAll(snapshot.getCommittingChangeSets());
     committedChangeSets.putAll(snapshot.getCommittedChangeSets());
   }
 
   /**
+   * Check if in-progress transactions need to be migrated to have expiration time and type, if so do the migration.
+   * This is required for backwards compatibility, when long running transactions were represented
+   * with expiration time -1. This can be removed when we stop supporting SnapshotCodec version 1.
+   */
+  private Map<Long, InProgressTx> txnBackwardsCompatCheck(Map<Long, InProgressTx> inProgress) {
+    for (Map.Entry<Long, InProgressTx> entry : inProgress.entrySet()) {
+      if (entry.getValue().getExpiration() < 0) {
+        InProgressTx compatTx =
+          new InProgressTx(entry.getValue().getVisibilityUpperBound(), getTxExpiration(defaultLongTimeout), 
```

Comment:
It doesn't seem right that this is going to give a different expiration time depending on when the snapshot is being restored.  Essentially the expiration is being extended/reset on snapshot restore.  We should base the system time for the expiration on the transaction write pointer (time in millis = write pointer / TxConstants.MAX_TX_PER_MS).